text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+- Feature Name: `gen-fn`
+- Start Date: 2023-10-10
+- RFC PR: [rust-lang/rfcs#3513](https://github.com/rust-lang/rfcs/pull/3513)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Reserve the `gen` keyword in the 2024 edition.
+
+Add `gen {}` blocks to the language. These implement `Iterator` by `yield`ing
+elements. This is simpler and more intuitive than creating a custom type and
+manually implementing `Iterator` for that type, which requires writing an
+explicit `Iterator::next` method body. This is a change similar to adding `async
+{}` blocks that implement `Future` instead of having to manually write futures
+and their state machines.
+
+Furthermore, add `gen fn` to the language. `gen fn foo(arg: X) -> Y` desugars to
```

Comment:
Given that there's more finality in the choice of syntax now than what was discussed earlier, I am concerned about this syntax, particularly using the return type to mean the yielded type. This is a one way door change that precludes extending `gen fn` to support non-() return types in the future. Given that `yield` is already a reserved keyword, I would prefer to use that instead of `->` (or use `yields` as a contextual keyword).

Reply:
It may make sense to just remove `gen fn` as part of the normative proposal in this document.  This doesn't need to be decided to reserve the keyword or even for an initial stabilization of an expression syntax.

There's also the issue that there has been some discussion around the idea that we may have erred on this point with `async fn`, i.e. maybe it should have included the `-> impl Future<..>` after all.  Eliding that has made a number of problems more difficult as bounds on the returned `Future` cannot be expressed.  It's something we should think about here also.

Anyway, we're planning some deeper revisions, and we'll definitely think about this point.

Reply:
(Apologies if this has been covered elsewhere in the discussion, I'm behind on the comments.)

I don't know if removing `gen fn` from part of the syntax helps much from a keyword standpoint because we'll still presumably want something like `gen { }` blocks, and that needs a keyword.

As far as syntax goes, right now I'm kind of fond of `gen fn foo() yield i32 { }`, or even just `fn foo() yield i32 { }`.

One reason I like this is that it generalizes pretty well to coroutines (at least for the ones where `resume` doesn't take an argument), since we could write `fn foo() -> bool yield i32 { }` for a coroutine that returns `bool` and yields some number of `i32`s along the way.

Reply:
I would like to see `gen fn` removed from the normative proposal as well, for now, just so that we can go ahead and approve the `gen` keyword and `gen`/`async gen` blocks.

Reply:
This was removed.